python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,182 @@
+        object_info = self._fs.info(self.location)
+        if size := object_info.get("Size"):
+            return size
+        elif size := object_info.get("size"):
+            return size
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._fs.lexists(self.location)
+
+    def open(self):
+        """Create an input stream for reading the contents of the file
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object
+        """
+        return self._fs.open(self.location, "rb")
+
+
+class FsspecOutputFile(OutputFile):
+    """An output file implementation for the FsspecFileIO
+
+    Args:
+        location(str): A URI to a file location
+        fs: An fsspec filesystem instance
+    """
+
+    def __init__(self, location: str, fs):
+        self._fs = fs
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._fs.info(self.location)
+        if size := object_info.get("Size"):
+            return size
+        elif size := object_info.get("size"):
+            return size
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._fs.lexists(self.location)
+
+    def create(self, overwrite: bool = False):
+        """Create an output stream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object
+
+        Raises:
+            FileExistsError: If the file already exists at the location and overwrite is set to False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._fs.touch(self.location)
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Returns a new FsspecInputFile for the location at `self.location`"""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, client_kwargs: Dict[str, str] = EMPTY_DICT, **fs_kwargs):
```

Comment:
I think this should be `def __init__(self, **properties: str):` That matches what we use for catalogs.

This is also where we want to pull out properties to configure the `FileIO`. Since the properties will differ between file systems, I think you may want to just pass this map to a method that instantiates the right file system:

```python
def _s3(properties: Properties) -> FileSystem:
    from s3fs import S3FileSystem

    # build client_kwargs from properties
    client_kwargs = ...

    return S3FileSystem(client_kwargs=client_kwargs, ...)

SCHEME_TO_FS = {
    "s3": _s3,
    "s3a": _s3,
    "s3n": _s3,
    ...
}
```

Reply:
Updated to use this, thanks @rdblue!